File: 프로그래머스/lv1/권채린/신고 결과 받기.cpp
#include <string>
#include <vector>
#include <map>
#include <set>

using namespace std;

vector<int> solution(vector<string> id_list, vector<string> report, int k) {
    vector<int> answer;
    map<string, set<string> report_list; // from, to
    map<string, int> id_count;  // 신고당한 횟수
    
    for (int i=0; i<report.size(); i++) {
        int blank = report[i].find(" ");
        string from = report[i].substr(0, blank);
        string to = report[i].substr(blank); // 공백 이후부터 to를 추출
        
        if (report_list[from].find(to) == report_list[from].end()) {
            report_list[from].insert(to);
            id_count[to]++;
        }
    }

    for (string s : id_list) {
        int mail=0;
        for(string str : report_list[s]) {
            if(id_count[str]>=k) {
                mail++;
            }
        }
        answer.push_back(mail);
        
    }

    return answer;
}

File: 프로그래머스/lv1/이재표/신고 결과 받기.java
import java.util.*;
class Solution {
    public int[] solution(String[] id_list, String[] reports, int k) {
        Map<String,Set<String>>report_people=new HashMap<>();
        Map<String,Set<String>>people=new HashMap<>();
        Map<String,Integer>mail=new HashMap<>();
        for(String id:id_list){
            people.put(id,new HashSet<>());
            report_people.put(id,new HashSet<>());
        }
        for(String report:reports){
            String[]r=report.split(" ");
            report_people.get(r[1]).add(r[0]);
            people.get(r[0]).add(r[1]);
        }
        List<String>upper_k=new ArrayList<>();
        for(String rp:report_people.keySet()){
            Integer cand=report_people.get(rp).size();
            if(cand>=k){
                upper_k.add(rp);
            }
        }
        for(String person:people.keySet()){
            for(String u:upper_k){
                if(people.get(person).contains(u)){
                    mail.put(person, mail.getOrDefault(person, 0) + 1);
                }
            }
        }
        int idx=0;
        int n=id_list.length;
        int[]answer=new int[n];
        for(int i=0;i<n;i++){
            answer[i] = mail.getOrDefault(id_list[i], 0);
        }
        return answer;
    }
}
